File: flex/src/com/intellij/lang/actionscript/parsing/ActionScriptParser.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.lang.actionscript.parsing;

import com.intellij.lang.PsiBuilder;
import com.intellij.lang.javascript.JSKeywordSets;
import com.intellij.lang.javascript.JSTokenTypes;
import com.intellij.lang.javascript.JavaScriptSupportLoader;
import com.intellij.lang.javascript.parsing.*;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

/**
 * @author Konstantin.Ulitin
 */
public final class ActionScriptParser extends JavaScriptParser<ExpressionParser, ActionScriptStatementParser, ActionScriptFunctionParser, JSPsiTypeParser> {
  public ActionScriptParser(PsiBuilder builder) {
    super(JavaScriptSupportLoader.ECMA_SCRIPT_L4, builder);
    myStatementParser = new ActionScriptStatementParser(this);
    myFunctionParser = new ActionScriptFunctionParser(this);
    myExpressionParser = new ActionScriptExpressionParser(this);
  }

  @Override
  public boolean isIdentifierToken(IElementType tokenType) {
    return JSKeywordSets.AS_IDENTIFIER_TOKENS_SET.contains(tokenType);
  }

  @Override
  public void buildTokenElement(@NotNull IElementType type) {
    final PsiBuilder.Marker marker = builder.mark();
    builder.advanceLexer();
    if (builder.getTokenType() == JSTokenTypes.GENERIC_SIGNATURE_START) {
      myTypeParser.parseECMA4GenericSignature();
    }
    marker.done(type);
  }

  @Override
  protected void doParseJS() {
    while (!builder.eof()) {
      if (builder.getTokenType() == JSTokenTypes.AT) {
        builder.advanceLexer();
        getStatementParser().parseAttributeBody();
      }
      else {
        getStatementParser().parseStatement();
      }
    }
  }
}

File: flex/src/com/intellij/lang/actionscript/parsing/ActionScriptParser.kt
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.lang.actionscript.parsing

import com.intellij.lang.PsiBuilder
import com.intellij.lang.javascript.JSKeywordSets
import com.intellij.lang.javascript.JSTokenTypes
import com.intellij.lang.javascript.parsing.JSParsingContextUtil
import com.intellij.lang.javascript.parsing.JavaScriptParserBase
import com.intellij.lang.javascript.parsing.modifiers.JSModifiersStructure
import com.intellij.lang.javascript.parsing.modifiers.JSModifiersStructure.JSModifiersParseResult
import com.intellij.psi.tree.IElementType
import java.util.*
import java.util.function.Predicate

/**
 * @author Konstantin.Ulitin
 */
class ActionScriptParser(internal val builder: PsiBuilder) {

  val expressionParser: ActionScriptExpressionParser =
    ActionScriptExpressionParser(this)

  val statementParser: ActionScriptStatementParser =
    ActionScriptStatementParser(this)

  val functionParser: ActionScriptFunctionParser =
    ActionScriptFunctionParser(this)

  internal val typeParser: ActionScriptPsiTypeParser by lazy {
    ActionScriptPsiTypeParser(this)
  }

  internal val xmlParser: ActionScriptXmlTokensParser by lazy {
    ActionScriptXmlTokensParser(this)
  }

  fun parseJS(root: IElementType) {
    val rootMarker = builder.mark()
    val forceContext = builder.getUserData(JavaScriptParserBase.FORCE_CONTEXT_KEY)
    if (forceContext != null) {
      if (forceContext == JavaScriptParserBase.ForceContext.Parameter && builder.tokenType === JSTokenTypes.DOT_DOT_DOT) {
        builder.advanceLexer()
      }
      else {
        if (forceContext == JavaScriptParserBase.ForceContext.TypeAllowEmpty && builder.eof()) {
          rootMarker.done(root)
          return
        }
        typeParser.parseType()
      }
    }
    else {
      builder.putUserData(JSParsingContextUtil.ASYNC_METHOD_KEY, JSParsingContextUtil.IS_TOP_LEVEL_ASYNC)
    }
    doParseJS()
    rootMarker.done(root)
  }

  fun isIdentifierName(firstToken: IElementType?): Boolean {
    return JSKeywordSets.IDENTIFIER_NAMES.contains(firstToken)
  }

  fun parseModifiers(
    structure: JSModifiersStructure,
    dropIfEmpty: Boolean,
    isPossibleStateAfterModifiers: Predicate<in PsiBuilder>,
  ): EnumSet<JSModifiersParseResult> {
    val offsetBefore = builder.currentOffset
    var attrList = builder.mark()

    var parseResults = structure.parseOptimistically(builder)

    if (parseResults.contains(JSModifiersParseResult.LEXER_ADVANCED) && !isPossibleStateAfterModifiers.test(builder)) {
      attrList.rollbackTo()
      attrList = builder.mark()
      parseResults = structure.parse(builder, isPossibleStateAfterModifiers)
    }

    val lexerAdvanced = builder.currentOffset > offsetBefore
    if (lexerAdvanced) {
      parseResults.add(JSModifiersParseResult.LEXER_ADVANCED)
    }
    else {
      parseResults.remove(JSModifiersParseResult.LEXER_ADVANCED)
    }
    if (dropIfEmpty && !lexerAdvanced) {
      attrList.drop()
    }
    else {
      attrList.done(functionParser.attributeListElementType)
    }

    return parseResults
  }

  fun isIdentifierToken(tokenType: IElementType?): Boolean {
    return JSKeywordSets.AS_IDENTIFIER_TOKENS_SET.contains(tokenType)
  }

  fun buildTokenElement(type: IElementType) {
    val marker = builder.mark()
    builder.advanceLexer()
    if (builder.tokenType === JSTokenTypes.GENERIC_SIGNATURE_START) {
      typeParser.parseECMA4GenericSignature()
    }
    marker.done(type)
  }

  fun doParseJS() {
    while (!builder.eof()) {
      if (builder.tokenType === JSTokenTypes.AT) {
        builder.advanceLexer()
        statementParser.parseAttributeBody()
      }
      else {
        statementParser.parseStatement()
      }
    }
  }
}
